src/server/clients_info/client_visibility.rs

```diff
@@ -0,0 +1,284 @@
+    pub(crate) fn read_entity_state(&mut self, entity: Entity) {
+        match &mut self.filter {
+            VisibilityFilter::All { just_connected } => {
+                if *just_connected {
+                    self.entity_state = EntityState::JustVisible
+                } else {
+                    self.entity_state = EntityState::Visible
+                }
+            }
+            VisibilityFilter::Blacklist { list, .. } => match list.get(&entity) {
+                Some(true) => self.entity_state = EntityState::JustVisible,
+                Some(false) => self.entity_state = EntityState::Hidden,
+                None => self.entity_state = EntityState::Visible,
+            },
+            VisibilityFilter::Whitelist { list, .. } => match list.get(&entity) {
+                Some(true) => self.entity_state = EntityState::JustVisible,
+                Some(false) => self.entity_state = EntityState::Visible,
+                None => self.entity_state = EntityState::Hidden,
+            },
+        }
+    }
+
+    /// Returns state obtained from last call of [`Self::read_entity_state`].
+    pub(crate) fn entity_state(&self) -> EntityState {
+        self.entity_state
+    }
+}
+
+/// Filter for [`ClientVisibility`] based on [`VisibilityPolicy`].
+enum VisibilityFilter {
+    All {
+        /// Indicates that the client has just connected to the server.
+        just_connected: bool,
+    },
+    Blacklist {
+        /// All blacklisted entities and an indicator of whether it is in the queue for deletion at this tick.
+        list: EntityHashMap<Entity, bool>,
+        /// All entities that were removed from the list on this tick.
+        added: EntityHashSet<Entity>,
+        /// All entities that were added to the list on this tick.
+        removed: EntityHashSet<Entity>,
+    },
+    Whitelist {
+        /// All whitelisted entities and an indicator whether it was added to the list on this tick.
+        list: EntityHashMap<Entity, bool>,
+        /// All entities that were added to the list on this tick.
+        added: EntityHashSet<Entity>,
+        /// All entities that were removed from the list on this tick.
+        removed: EntityHashSet<Entity>,
+    },
+}
+
+/// Visibility state for an entity.
+#[derive(PartialEq, Default, Clone, Copy)]
+pub(crate) enum EntityState {
+    #[default]
+    None,
+    Hidden,
+    Visible,
+    JustVisible,
+}
+
+enum VisibilityLostIter<T> {
+    AllVisible,
+    Lost(T),
+}
+
+impl<T: Iterator> Iterator for VisibilityLostIter<T> {
+    type Item = T::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            VisibilityLostIter::AllVisible => None,
+            VisibilityLostIter::Lost(entities) => entities.next(),
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        match self {
+            VisibilityLostIter::AllVisible => (0, Some(0)),
+            VisibilityLostIter::Lost(entities) => entities.size_hint(),
+        }
+    }
```

Comment:
Looks like it's not triggered in our code at all since we never do things like `.collect()` or `.rev()`. Keep anyway?

Reply:
Even if we use `Drain`, `size_hint` will still not be used.